### PR TITLE
Avoid mutation of `os.environ`

### PR DESCRIPTION
**Pull request by asi1024**
This PR is a code-fix of a test to use `mock.patch` to `cupy.cuda.compiler.*` instead of mutation of `os.environ`.

### CODE REVIEW

File: tests/cupy_tests/core_tests/test_raw.py [hunk 1]
```diff
@@ -327,13 +329,16 @@
...
-if 'CUPY_CACHE_DIR' in os.environ:
-    _old_cache_dir = os.environ['CUPY_CACHE_DIR']
-    _is_cache_env_var_set = True
-else:
-    _old_cache_dir = os.path.expanduser('~/.cupy/kernel_cache')
-    _is_cache_env_var_set = False
-_test_cache_dir = None
+
+@contextlib.contextmanager
+def use_temporary_cache_dir():
```
Comment (asi1024) [inline tests/cupy_tests/core_tests/test_raw.py:334]: Is `mock_temporary_cache_dir` better?

Comment (leofang) [reply tests/cupy_tests/core_tests/test_raw.py:334]: I am fine either way.

**Comment by leofang** [comment]
Kicking off the CI to expedite...

Jenkins, test this please

**Comment by leofang** [comment]
Failed tests are irrelevant and should be fixed by #3750. Rerun to confirm...

Jenkins, test this please

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by kmaehashi** [comment]
Thanks @asi1024 and @leofang!

We talked that it is better to provide an internal method to change cache dir because mocking with function name string is fragile.

**Comment by leofang** [comment]
Thanks! @kmaehashi Could you elaborate on the mocking? 😅 What could go wrong with it?

**Pull request merged**